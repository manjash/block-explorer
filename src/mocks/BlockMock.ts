import Unusual from 'unusual'
import Block from '../types/Block'
import pkg from '../../package.json'

const u = new Unusual(pkg.version)

const getRandomIntInclusive = (min: number, max: number) => u.integer({ min, max })
const alphabet = 'abcdefghijklmnopqrstuvwxyz'.split('')
const word = (length: number): string => Array.from({ length }, () => u.pick(alphabet)).join('')

export const generateBlock = (): Block => ({
  id: getRandomIntInclusive(0, 10000),
  sequence: getRandomIntInclusive(10000, 99999),
  hash: `${getRandomIntInclusive(
    0,
    10000,
  )}25a9a4efa997f70c01efd06a3b27d1e5a2346ab1ff9b6147e816de`,
  previous_block_hash: `${getRandomIntInclusive(
    0,
    10000,
  )}25a9a4efa997f70c01efd06a3b27d1e5a2346ab1ff9b6147e816de`,
  main: true,
  difficulty: getRandomIntInclusive(10000, 99999),
  size: getRandomIntInclusive(10000, 99999),
  transactions: [],
  transactionsCount: 0,
  timestamp: new Date(
    2022,
    0,
    1,
    u.integer({ min: 0, max: 23 }),
    u.integer({ min: 0, max: 59 }),
  ),
  graffiti: word(u.integer({ min: 1, max: 16 })),
})

export const generateBlocksWithMatchingHashes = (total: number): Block[] => {
  const list = blockList(total)
  return list
    .sort(({ sequence: a, sequence: b }) => a - b)
    .reduce((agg: Block[], block: Block): Block[] => {
      const prev = agg[agg.length - 1]
      if (prev) {
        const copy = { ...block, previous_block_hash: prev.hash }
        return agg.concat(copy)
      }
      return agg.concat(block)
    }, [])
}

export const blockList = (numberOfBlocks: number): Array<Block> =>
  Array.from({ length: numberOfBlocks }, generateBlock)
