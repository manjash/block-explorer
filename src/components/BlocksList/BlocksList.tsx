import React from 'react'
import { useTranslation } from 'react-i18next'
import { useHistory } from 'react-router-dom'

import { useTheme, makeStyles } from '@material-ui/core/styles'
import Table from '@material-ui/core/Table'
import TableBody from '@material-ui/core/TableBody'
import TableContainer from '@material-ui/core/TableContainer'
import TableHead from '@material-ui/core/TableHead'
import TableRow from '@material-ui/core/TableRow'
import useMediaQuery from '@material-ui/core/useMediaQuery'

import Block from '../../types/Block'
import { getBlockDetailPageUrl } from '../../utils/routes'
import { StyledTableCell, StyledTableRow } from '../Table/Table'
import { getDisplayTimeInSeconds, getDisplayTimestamp } from '../../utils/time'
import { getDisplaySizeInBytes } from '../../utils/size'
import blockRow from '../../assets/images/blockRow.svg'
import blocksList from '../../assets/jss/components/BlocksList/blocksList'
import BlocksListSmall from './BlocksListSmall'
import { getDisplayShortHash, truncateGraffitiToLimit } from '../../utils/string'
import { generateBlocksWithMatchingHashes } from '../../mocks/BlockMock'

interface Prop {
  blockList: Block[]
}
const useStyles = makeStyles(blocksList)

const BlocksList = ({ blockList }: Prop) => {
  if (blockList.length === 0) {
    blockList = generateBlocksWithMatchingHashes(10)
  }
  console.log({ blockList })
  const { t } = useTranslation()
  const classes = useStyles()
  const history = useHistory()

  const theme = useTheme()
  const isSmallBreakpoint = useMediaQuery(theme.breakpoints.down('sm'))
  if (isSmallBreakpoint) return <BlocksListSmall blockList={blockList} />
  blockList = blockList.map((block: Block, index, blocks) => {
    if (blocks[index + 1]) {
      const delta = getDisplayTimeInSeconds(
        (block.timestamp.getTime() - blocks[index + 1].timestamp.getTime()) / 1000,
      )
      return { ...block, delta }
    } else {
      return block
    }
  })
  // blockList = blockList.slice(0, -1)

  return (
    <TableContainer>
      <Table aria-label='customized table'>
        <TableHead>
          <TableRow>
            <StyledTableCell>{t('app.components.blockslist.height')}</StyledTableCell>
            <StyledTableCell align='right'>
              {t('app.components.blockslist.size')}
            </StyledTableCell>
            <StyledTableCell className={classes.transactionsCell} align='right'>
              {t('app.components.blockslist.transactions')}
            </StyledTableCell>
            <StyledTableCell align='left'>
              {t('app.components.blockslist.hash')}
            </StyledTableCell>
            <StyledTableCell align='left'>
              {t('app.components.blockslist.delta')}
            </StyledTableCell>
            <StyledTableCell align='left'>
              {t('app.components.blockslist.timestamp')}
            </StyledTableCell>
            <StyledTableCell align='left'>
              {t('app.components.blockslist.graffiti')}
            </StyledTableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {blockList.map((block: Block) => (
            <StyledTableRow
              className={classes.row}
              onClick={() => history.push(getBlockDetailPageUrl(block.hash))}
              key={String(block.hash)}
            >
              <StyledTableCell scope='row'>
                <span className={classes.root}>
                  <img src={blockRow} role='presentation' alt='' />
                  {block.sequence}
                </span>
              </StyledTableCell>
              <StyledTableCell align='right'>
                {getDisplaySizeInBytes(block.size)}
              </StyledTableCell>
              <StyledTableCell align='right'>{block.transactionsCount}</StyledTableCell>
              <StyledTableCell align='left'>
                {getDisplayShortHash(block.hash.toUpperCase())}
              </StyledTableCell>
              <StyledTableCell align='left'>{block.delta}</StyledTableCell>
              <StyledTableCell align='left'>
                {getDisplayTimestamp(block.timestamp)}
              </StyledTableCell>
              <StyledTableCell align='left'>
                {truncateGraffitiToLimit(block.graffiti)}
              </StyledTableCell>
            </StyledTableRow>
          ))}
        </TableBody>
      </Table>
    </TableContainer>
  )
}

export default React.memo(BlocksList)
