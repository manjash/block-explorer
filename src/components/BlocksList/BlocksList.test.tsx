import React from 'react'
import { shallow } from 'enzyme'
import snap from 'enzyme-to-json'

import TableRow from '@material-ui/core/TableRow'

import { generateBlocksWithMatchingHashes as mockData } from '../../mocks/BlockMock'
import { StyledTableRow } from '../../components/Table/Table'

import BlocksList from './BlocksList'

const CLEANUP = /(WithStyles\()+ForwardRef\((\w+)\)\)\)/g
// const snap = (x: any) =>
//   ((json(x) as unknown) as string)
//    .split('\n')
//    .map((z: string) => z.replace(CLEANUP, '$2'))
//    .join('\n')

describe('<BlocksList />', () => {
  test('It renders correctly', () => {
    const data = mockData(3)
    // console.log({ data })
    const wrapper = shallow(<BlocksList blockList={data} />)
    expect(snap(wrapper)).toMatchSnapshot()
    expect(wrapper.find(TableRow)).toHaveLength(1)
    expect(wrapper.find(StyledTableRow)).toHaveLength(30)
  })
})
