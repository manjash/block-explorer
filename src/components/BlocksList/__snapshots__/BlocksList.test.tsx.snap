// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BlocksList /> It renders correctly 1`] = `
<WithStyles(ForwardRef(TableContainer))>
  <WithStyles(ForwardRef(Table))
    aria-label="customized table"
  >
    <WithStyles(ForwardRef(TableHead))>
      <WithStyles(ForwardRef(TableRow))>
        <WithStyles(WithStyles(ForwardRef(TableCell)))>
          app.components.blockslist.height
        </WithStyles(WithStyles(ForwardRef(TableCell)))>
        <WithStyles(WithStyles(ForwardRef(TableCell)))
          align="right"
        >
          app.components.blockslist.size
        </WithStyles(WithStyles(ForwardRef(TableCell)))>
        <WithStyles(WithStyles(ForwardRef(TableCell)))
          align="right"
          className="makeStyles-transactionsCell-3"
        >
          app.components.blockslist.transactions
        </WithStyles(WithStyles(ForwardRef(TableCell)))>
        <WithStyles(WithStyles(ForwardRef(TableCell)))
          align="left"
        >
          app.components.blockslist.hash
        </WithStyles(WithStyles(ForwardRef(TableCell)))>
        <WithStyles(WithStyles(ForwardRef(TableCell)))
          align="left"
        >
          app.components.blockslist.delta
        </WithStyles(WithStyles(ForwardRef(TableCell)))>
        <WithStyles(WithStyles(ForwardRef(TableCell)))
          align="left"
        >
          app.components.blockslist.timestamp
        </WithStyles(WithStyles(ForwardRef(TableCell)))>
        <WithStyles(WithStyles(ForwardRef(TableCell)))
          align="left"
        >
          app.components.blockslist.graffiti
        </WithStyles(WithStyles(ForwardRef(TableCell)))>
      </WithStyles(ForwardRef(TableRow))>
    </WithStyles(ForwardRef(TableHead))>
    <WithStyles(ForwardRef(TableBody))>
      <WithStyles(WithStyles(ForwardRef(TableRow)))
        className="makeStyles-row-1"
        key="444225a9a4efa997f70c01efd06a3b27d1e5a2346ab1ff9b6147e816de"
        onClick={[Function]}
      >
        <WithStyles(WithStyles(ForwardRef(TableCell)))
          scope="row"
        >
          <span
            className="makeStyles-root-2"
          >
            <img
              alt=""
              role="presentation"
              src="blockRow.svg"
            />
            25561
          </span>
        </WithStyles(WithStyles(ForwardRef(TableCell)))>
        <WithStyles(WithStyles(ForwardRef(TableCell)))
          align="right"
        >
          app.units.size.bytes30.38,kb
        </WithStyles(WithStyles(ForwardRef(TableCell)))>
        <WithStyles(WithStyles(ForwardRef(TableCell)))
          align="right"
        >
          0
        </WithStyles(WithStyles(ForwardRef(TableCell)))>
        <WithStyles(WithStyles(ForwardRef(TableCell)))
          align="left"
        >
          4442...16DE
        </WithStyles(WithStyles(ForwardRef(TableCell)))>
        <WithStyles(WithStyles(ForwardRef(TableCell)))
          align="left"
        >
          app.units.time.seconds3240
        </WithStyles(WithStyles(ForwardRef(TableCell)))>
        <WithStyles(WithStyles(ForwardRef(TableCell)))
          align="left"
        >
          1/1/2022 9:03:00PM
        </WithStyles(WithStyles(ForwardRef(TableCell)))>
        <WithStyles(WithStyles(ForwardRef(TableCell)))
          align="left"
        >
          khpigcoxyzh
        </WithStyles(WithStyles(ForwardRef(TableCell)))>
      </WithStyles(WithStyles(ForwardRef(TableRow)))>
      <WithStyles(WithStyles(ForwardRef(TableRow)))
        className="makeStyles-row-1"
        key="972925a9a4efa997f70c01efd06a3b27d1e5a2346ab1ff9b6147e816de"
        onClick={[Function]}
      >
        <WithStyles(WithStyles(ForwardRef(TableCell)))
          scope="row"
        >
          <span
            className="makeStyles-root-2"
          >
            <img
              alt=""
              role="presentation"
              src="blockRow.svg"
            />
            42860
          </span>
        </WithStyles(WithStyles(ForwardRef(TableCell)))>
        <WithStyles(WithStyles(ForwardRef(TableCell)))
          align="right"
        >
          app.units.size.bytes93.29,kb
        </WithStyles(WithStyles(ForwardRef(TableCell)))>
        <WithStyles(WithStyles(ForwardRef(TableCell)))
          align="right"
        >
          0
        </WithStyles(WithStyles(ForwardRef(TableCell)))>
        <WithStyles(WithStyles(ForwardRef(TableCell)))
          align="left"
        >
          9729...16DE
        </WithStyles(WithStyles(ForwardRef(TableCell)))>
        <WithStyles(WithStyles(ForwardRef(TableCell)))
          align="left"
        >
          app.units.time.seconds71520
        </WithStyles(WithStyles(ForwardRef(TableCell)))>
        <WithStyles(WithStyles(ForwardRef(TableCell)))
          align="left"
        >
          1/1/2022 8:09:00PM
        </WithStyles(WithStyles(ForwardRef(TableCell)))>
        <WithStyles(WithStyles(ForwardRef(TableCell)))
          align="left"
        >
          lwsag
        </WithStyles(WithStyles(ForwardRef(TableCell)))>
      </WithStyles(WithStyles(ForwardRef(TableRow)))>
      <WithStyles(WithStyles(ForwardRef(TableRow)))
        className="makeStyles-row-1"
        key="535425a9a4efa997f70c01efd06a3b27d1e5a2346ab1ff9b6147e816de"
        onClick={[Function]}
      >
        <WithStyles(WithStyles(ForwardRef(TableCell)))
          scope="row"
        >
          <span
            className="makeStyles-root-2"
          >
            <img
              alt=""
              role="presentation"
              src="blockRow.svg"
            />
            35521
          </span>
        </WithStyles(WithStyles(ForwardRef(TableCell)))>
        <WithStyles(WithStyles(ForwardRef(TableCell)))
          align="right"
        >
          app.units.size.bytes72.01,kb
        </WithStyles(WithStyles(ForwardRef(TableCell)))>
        <WithStyles(WithStyles(ForwardRef(TableCell)))
          align="right"
        >
          0
        </WithStyles(WithStyles(ForwardRef(TableCell)))>
        <WithStyles(WithStyles(ForwardRef(TableCell)))
          align="left"
        >
          5354...16DE
        </WithStyles(WithStyles(ForwardRef(TableCell)))>
        <WithStyles(WithStyles(ForwardRef(TableCell)))
          align="left"
        />
        <WithStyles(WithStyles(ForwardRef(TableCell)))
          align="left"
        >
          1/1/2022 12:17:00AM
        </WithStyles(WithStyles(ForwardRef(TableCell)))>
        <WithStyles(WithStyles(ForwardRef(TableCell)))
          align="left"
        >
          l
        </WithStyles(WithStyles(ForwardRef(TableCell)))>
      </WithStyles(WithStyles(ForwardRef(TableRow)))>
    </WithStyles(ForwardRef(TableBody))>
  </WithStyles(ForwardRef(Table))>
</WithStyles(ForwardRef(TableContainer))>
`;
